infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,28 +73,40 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheHash.get(conflictKey);
+    }
+
+    private long firstAddress() {
+        if (cacheEntries.peek() == null)
+            return Address.NOT_FOUND;
+        return cacheEntries.peek().txVersion;
     }
 
     /**
-     * Invalidate all records up to a trim mark.
+     * Invalidate the first record with the minAddress.
+     */
+    private void invalidateFirst() {
+        ConflictTxStreamEntry entry = cacheEntries.poll();
+        cacheHash.remove(entry.conflictTxStream);
+        maxConflictWildcard = Math.max(entry.txVersion, maxConflictWildcard);
+        log.trace("Updating maxConflictWildcard. Old = '{}', new ='{}' conflictParam = '{}'.",
+                maxConflictWildcard, entry.txVersion, entry.conflictTxStream);
+    }
+
+    /**
+     * Invalidate all records up to a trim mark (not included).
      *
      * @param trimMark trim mark
-     */
+     * */
     public void invalidateUpTo(long trimMark) {
         log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
-
         AtomicLong entries = new AtomicLong();
+        long first;
 
-        conflictCache.asMap().forEach((key, txVersion) -> {
-            if (txVersion == null || txVersion >= trimMark) {
-                return;
-            }
-
-            conflictCache.invalidate(key);
+        while((first = firstAddress()) != Address.NOT_FOUND && first < trimMark) {
```

Comment:
You can do this instead.

https://github.com/CorfuDB/CorfuDB/pull/2300/files#r364366983